netket/sampler/autoreg.py

```diff
@@ -70,43 +66,51 @@ def is_exact(sampler):
     def _init_cache(sampler, model, σ, key):
         variables = model.init(key, σ)
         if "cache" in variables:
-            _, cache = variables.pop("cache")
+            cache = variables["cache"]
         else:
             cache = None
         return cache
 
-    def _init_state(sampler, model, params, key):
-        new_key, key = jax.random.split(key)
-        σ = jnp.zeros((sampler.n_chains, sampler.hilbert.size), dtype=sampler.dtype)
-        cache = sampler._init_cache(model, σ, key)
-        return ARDirectSamplerState(σ=σ, cache=cache, key=new_key)
+    def _init_state(sampler, model, variables, key):
+        return ARDirectSamplerState(key=key)
 
-    def _reset(sampler, model, params, state):
+    def _reset(sampler, model, variables, state):
         return state
 
-    def _sample_chain(sampler, model, params, state, chain_length):
-        return _sample_chain(sampler, model, params, state, chain_length)
+    def _sample_chain(sampler, model, variables, state, chain_length):
+        return _sample_chain(sampler, model, variables, state, chain_length)
 
-    def _sample_next(sampler, model, params, state):
-        σ, new_state = sampler._sample_chain(model, params, state, 1)
+    def _sample_next(sampler, model, variables, state):
+        σ, new_state = sampler._sample_chain(model, variables, state, 1)
         σ = σ.squeeze(axis=0)
         return new_state, σ
 
 
 @partial(jax.jit, static_argnums=(1, 4))
-def _sample_chain(sampler, model, params, state, chain_length):
+def _sample_chain(sampler, model, variables, state, chain_length):
+    params = variables["params"]
+
     def scan_fun(carry, index):
         σ, cache, key = carry
+        if cache:
+            variables = freeze({"params": params, "cache": cache})
+        else:
+            variables = freeze({"params": params})
```

Comment:
you should assume that there might be more stuff other than cache, in the future, if someone is doing something weird.